File: day05/main.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"log"
)

const (
	puzzleInput = "input.txt"
)

func byteUpper(a byte) byte {
	if a-byte('a') > byte('z') {
		return a
	}
	return a - byte('a') + byte('A')
}

func react(a, b byte) bool {
	i := a
	j := b
	if i-byte('a') <= byte('z') {
		i -= byte('a') - byte('A')
	}
	if j-byte('a') <= byte('z') {
		j -= byte('a') - byte('A')
	}

	return a != b && i == j
}

type (
	Stack []byte
)

func (s *Stack) Push(a byte) {
	*s = append(*s, a)
}

func (s *Stack) Pop() byte {
	a := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return a
}

func (s *Stack) Peek() byte {
	return (*s)[len(*s)-1]
}

func (s *Stack) Len() int {
	return len(*s)
}

func reduce(line []byte) []byte {
	reduction := make(Stack, 0, len(line))
	for _, i := range line {
		if reduction.Len() == 0 {
			reduction.Push(i)
			continue
		}
		if react(reduction.Peek(), i) {
			reduction.Pop()
			continue
		}
		reduction.Push(i)
	}
	return []byte(reduction)
}

func filter(line []byte, c byte) int {
	b := make([]byte, 0, len(line))
	for _, i := range line {
		if i != c && byteUpper(i) != c {
			b = append(b, i)
		}
	}
	return len(reduce(b))
}

func main() {
	line, err := ioutil.ReadFile(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	line = bytes.TrimRight(line, "\n")

	fmt.Println(len(reduce(line)))

	shortestLength := 99999999
	for i := byte('A'); i <= byte('Z'); i++ {
		if size := filter(line, i); size < shortestLength {
			shortestLength = size
		}
	}

	fmt.Println(shortestLength)
}

File: day05/src/main.rs
use std::fs::File;
use std::io::prelude::*;

const PUZZLEINPUT: &str = "input.txt";

fn is_reactive(a: &char, b: &char) -> bool {
    a != b && a.to_ascii_uppercase() == b.to_ascii_uppercase()
}

fn reduce(state: &Vec<char>) -> Vec<char> {
    let mut acc = Vec::new();
    for i in state.iter() {
        if {
            match acc.last() {
                Some(x) => is_reactive(i, x),
                None => false,
            }
        } {
            acc.pop();
        } else {
            acc.push(*i);
        }
    }
    acc
}

fn is_not_char(a: &char, b: &char) -> bool {
    a != b && a.to_ascii_uppercase() != b.to_ascii_uppercase()
}

fn main() {
    let file = File::open(PUZZLEINPUT).expect("Failed to open file");
    let init_state = file
        .bytes()
        .map(|x| x.expect("Failed to read byte") as char)
        .filter(char::is_ascii_alphanumeric)
        .collect::<Vec<_>>();

    println!("{}", reduce(&init_state).len());

    let mut max = 99999999;
    for i in ('A' as u8)..=('Z' as u8) {
        let c = i as char;
        let l = reduce(
            &init_state
                .iter()
                .filter(|x| is_not_char(x, &c))
                .map(|&x| x)
                .collect::<Vec<_>>(),
        )
        .len();
        if l < max {
            max = l;
        }
    }
    println!("{}", max);
}
